### PR TITLE
Update whitening

### PR DESCRIPTION
**Pull request by EthanMarx**
This PR addresses issue #81. The `WhiteningTransform` object now takes in Optional parameters `highpass` and `fduration`.

### CODE REVIEW

File: libs/data/bbhnet/data/transforms/whitening.py [hunk 1]
```diff
@@ -88,17 +113,75 @@ def forward(self, X: torch.Tensor) -> torch.Tensor:
         X = X.transpose(2, 0)
         X = X - X.mean(axis=0)
         X = X.transpose(0, 2)
-        X *= self.window
+
+        self.pad = int(np.ceil(self.time_domain_filter.size(-1) / 2))
```
Comment (alecgunny) [inline libs/data/bbhnet/data/transforms/whitening.py:117]: Probably worth either assigning as an attribute up front, or just using as a variable in this namespace (i.e. no `self.`, just `pad`)

File: libs/data/bbhnet/data/transforms/whitening.py [hunk 2]
```diff
@@ -88,17 +113,75 @@ def forward(self, X: torch.Tensor) -> torch.Tensor:
...
+            conv[:, :, : nfft - self.pad] = torch.nn.functional.conv1d(
+                X[:, :, :nfft],
+                self.time_domain_filter,
+                groups=self.num_ifos,
+                padding="same",
+            )[:, :, : nfft - self.pad]
+
+            # process chunks of length nstep
+            k = nfft - self.pad
+            while k < self.kernel_size - nfft + self.pad:
```
Comment (alecgunny) [inline libs/data/bbhnet/data/transforms/whitening.py:163]: Looping and doing multiple convolutions is going to be pretty inefficient on GPU, especially since each convolution will be comparatively small because it only operates on a subslice (basically you're going to waste time launching a bunch of kernels for ops that will be pretty small, while GPUs can parallelize individual operations really efficiently). Consider using the torch [unfold](https://pytorch.org/docs/stable/generated/torch.nn.Unfold.html) operation up front, which will basically create multiple windowed views of the timeseries at a given stride (which for reasons having to do with what this op is intended for, I believe you'll want to pass as the `dilation` argument, _not_ the `stride` argument).

Then you can do a single convolution along the last axis, slice off the pads so that the views are non-overlapping, then do a simple reshape. This solution might only work for timeseries whose length can be divided evenly into chunks like this, but it's probably worth enforcing this for efficiency, especially at inference time.

**Review by alecgunny — COMMENTED** [review]
Looks really good, curious to hear your thoughts on a more efficient implementation. The response could even just be that we don't plan on using `fduration`s that are short enough to trigger this implementation for the time being, and so we don't need to worry about efficiency in this PR and can open it as an issue for later.

**Comment by alecgunny** [comment]
Also the export tests take forever, partly because there's probably more permutations than is really necessary and partly because there's a bunch of permutations that are intended to fail (if the kernel length is less than the stride length) that still have to go to all the trouble of building and saving an NN that they will never use. If this is slowing you down, feel free to take away some of these permutations or I can also open up a quick PR to address the NN issue as well

File: libs/data/bbhnet/data/transforms/whitening.py (hunk 2 above)
Comment (alecgunny) [reply libs/data/bbhnet/data/transforms/whitening.py:166]: Yeah that sounds great

**Comment by EthanMarx** [comment]
>Also the export tests take forever, partly because there's probably more permutations than is really necessary and partly because there's a bunch of permutations that are intended to fail (if the kernel length is less than the stride length) that still have to go to all the trouble of building and saving an NN that they will never use. If this is slowing you down, feel free to take away some of these permutations or I can also open up a quick PR to address the NN issue as well




> Also the export tests take forever, partly because there's probably more permutations than is really necessary and partly because there's a bunch of permutations that are intended to fail (if the kernel length is less than the stride length) that still have to go to all the trouble of building and saving an NN that they will never use. If this is slowing you down, feel free to take away some of these permutations or I can also open up a quick PR to address the NN issue as well

Oh wow yeah they've been taking a while. Are there specific permutations less important than others?

**Comment by alecgunny** [comment]
Yeah now that I'm looking at these with fresh eyes I can definitely pare these down pretty substantially. Now that I have inference working I'll assign this as an issue to myself and get it dealt with tomorrow, should be pretty quick. Let's go ahead and merge this as-is.

**Review by alecgunny — APPROVED** [review]
